cedar-lean/Cedar/Partial/Response.lean

```diff
@@ -195,4 +203,30 @@ def Response.underapproximateDeterminingPolicies (resp : Partial.Response) : Set
     -- when the decision is Unknown, nothing is guaranteed to be determining.
     Set.empty
 
+/--
+  Re-evaluate with the given substitution for unknowns, giving a new
+  `Partial.Response`.
+
+  It's fine for some unknowns to not be in `subsmap`, in which case the returned
+  `Partial.Response` will still contain some (nontrivial) residuals.
+
+  Returns `none` if:
+    - the substitution is invalid (e.g., if trying to substitute a
+        non-`EntityUID` into `UidOrUnknown`)
+-/
+def Response.reEvaluateWithSubst (subsmap : Map String Partial.Value) : Partial.Response → Option Partial.Response
+  | { residuals, req, entities } => do
+  let req' ← req.subst subsmap
+  some {
+    residuals := residuals.filterMap λ residual => match residual with
+      | .error id e => some (.error id e)
+      | .residual id effect cond => match Partial.evaluate (cond.subst subsmap) req' (entities.subst subsmap) with
+        | .ok (.value (.prim (.bool false))) => none
+        | .ok (.value v) => some (.residual id effect v.asPartialExpr)
+        | .ok (.residual r) => some (.residual id effect r)
+        | .error e => some (.error id e)
```

Comment:
Might be clearer to move this out to a separate function & explicitly document that any residuals that evaluate to false are dropped.

Reply:
I can make a note, but I also note that this is the standard behavior of `Partial.isAuthorized`, and indeed documented as an invariant on `Partial.Response`.

Reply:
Which part exactly do you suggest moving out into a separate function?

Reply:
Aha, makes sense

Reply:
Factored this out in the latest commit